Add support for pages deployment auto pagination

Cyb3r-Jak3 opened the pull request:
## Description

Adds support for auto paginate for pages deployments. Support forcing deletion of a pages deployment.
Also removed an unused type and var.

## Has your change been tested?

Updated existing tests and they pass.

## Screenshots (if appropriate):

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.
- [X] This change is using publicly documented (api.cloudflare.com or developers.cloudflare.com) and stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/

Cyb3r-Jak3 committed:
Add support for pages deployment auto pagination

Cyb3r-Jak3 committed:
Change DeletePagesDeploymentParams to be the URL parameters

Cyb3r-Jak3 committed:
Fix linting

Cyb3r-Jak3 committed:
Pages support max of 25 for per_page

jacobbednarz commented on pages_deployment.go:
```diff
@@ -217,7 +230,7 @@ func (api *API) GetPagesDeploymentLogs(ctx context.Context, rc *ResourceContaine
 //
 // API reference: https://api.cloudflare.com/#pages-deployment-delete-deployment
-func (api *API) DeletePagesDeployment(ctx context.Context, rc *ResourceContainer, projectName, deploymentID string) error {
+func (api *API) DeletePagesDeployment(ctx context.Context, rc *ResourceContainer, projectName, deploymentID string, params DeletePagesDeploymentParams) error {
```
this should have a maximum of three params - the last three should be condensed into one.

Cyb3r-Jak3 committed:
PR comments

jacobbednarz committed:
update more method signatures to match conventions

jacobbednarz commented on pages_project_test.go:
```diff
@@ -583,9 +613,12 @@ func TestUpdatePagesProject(t *testing.T) {
 	mux.HandleFunc("/accounts/"+testAccountID+"/pages/projects/Test Pages Project", handler)
 
-	_, err := client.UpdatePagesProject(context.Background(), testAccountID, "Test Pages Project", *updateAttributes)
+	_, err := client.UpdatePagesProject(context.Background(), AccountIdentifier(""), *updateAttributes)
```
this test will be broken at the moment while i chat with @WalshyDev about what we are doing with friendly URL IDs since these shouldn't be able to change but i think we currently allow them to.

Cyb3r-Jak3 committed:
Add support for providing branch name when creating a deployment

Cyb3r-Jak3 committed:
Implement ResultInfo.DoneCount() for result_info that doesn't implement total_pages

jacobbednarz commented on pagination.go:
```diff
@@ -17,3 +17,10 @@ func (p ResultInfo) Next() ResultInfo {
+// DoneCount returns true for the last page and false otherwise.
+// Unlike Done(), this function uses the Total and PerPage fields to determine if
+// there are more pages to fetch.
+func (p ResultInfo) DoneCount() bool {
```
we shouldn't need this; if this is needed, we aren't sending the correct v4 response envelope and we should have the team address it.

Cyb3r-Jak3 replied:
I agree it shouldn't be needed. Unfortunately, there are other APIs besides deployments that don't return the full result_info information as well.

jacobbednarz replied:
can you shoot me a list please? this is breaking an API contract if it is on a listing endpoint.

Cyb3r-Jak3 replied:
I've updated this to use the correct way. Once pages backend has support, then this will be all set.

Cyb3r-Jak3 committed:
Merge branch 'master' into pages-deployment-autoPaginate

Cyb3r-Jak3 committed:
Update pages_deployment to use the correct way to pagination.

jacobbednarz committed:
Merge branch 'master' into pages-deployment-autoPaginate

Cyb3r-Jak3 commented:
I don't know what the plan is for the project name test, I know that the API will reject it if you try. Also, `total_pages` is now in the API response for listing deployments.

jacobbednarz committed:
split usage of `ID` and `Name`

Pages, unfortunately, uses a friendly name for the identifier in the URL
so to be able to update the `Name` value we need a way to differentiate
the two. The way I've solved this here is that `ID` is used for
addressing the resource via the UI or a stable identifier whereas `Name`
is used for updating the value.

jacobbednarz committed:
Merge branch 'master' into pages-deployment-autoPaginate

jacobbednarz committed:
fix test assertions
